src/main/java/io/supertokens/jwt/JWTSigningKey.java

```diff
@@ -0,0 +1,247 @@
+     * @throws StorageQueryException If there is an error interacting with the database
+     * @throws StorageTransactionLogicException If there is an error interacting with the database
+     */
+    public List<JWTSigningKeyInfo> getAllSigningKeys()
+            throws StorageQueryException, StorageTransactionLogicException {
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+
+        if (storage.getType() == STORAGE_TYPE.SQL) {
+            SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
+
+            return sqlStorage.startTransaction(con -> {
+               List<JWTSigningKeyInfo> keys;
+               List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
+
+               if (keysFromStorage == null) {
+                   keys = new ArrayList<>();
+               } else {
+                   keys = keysFromStorage;
+               }
+
+               sqlStorage.commitTransaction(con);
+               return keys;
+            });
+        } else if (storage.getType() == STORAGE_TYPE.NOSQL_1) {
+            SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
+
+            List<JWTSigningKeyInfo> keys;
+            List<JWTSigningKeyInfo> keysFromStorage = noSQLStorage.getJWTSigningKeys_Transaction();
+
+            if (keysFromStorage == null) {
+                keys = new ArrayList<>();
+            } else {
+                keys = keysFromStorage;
+            }
+
+            return keys;
+        }
+
+        throw new QuitProgramException("Unsupported storage type detected");
+    }
+
+    /**
+     * Used to retrieve a key for JWT validation, for a given signing algorithm. If there are no keys in storage that match the given algorithm a new one is generated.
+     * @param algorithm The signing algorithm
+     * @return {@link JWTSigningKeyInfo} key
+     * @throws UnsupportedJWTSigningAlgorithmException If there is an error in the provided algorithm when getting/generating keys
+     * @throws StorageQueryException If there is an error interacting with the database
+     * @throws StorageTransactionLogicException If there is an error interacting with the database
+     * @throws DuplicateKeyIdException If there is an error setting generated keys in storage
+     */
+    public JWTSigningKeyInfo getOrCreateAndGetKeyForAlgorithm(SupportedAlgorithms algorithm)
+            throws UnsupportedJWTSigningAlgorithmException, StorageQueryException, StorageTransactionLogicException, DuplicateKeyIdException {
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+
+        if (storage.getType() == STORAGE_TYPE.SQL) {
+            SessionSQLStorage sqlStorage = (SessionSQLStorage) storage;
+
+            try {
+                return sqlStorage.startTransaction(con -> {
+                    JWTSigningKeyInfo keyInfo = null;
+
+                    List<JWTSigningKeyInfo> keysFromStorage = sqlStorage.getJWTSigningKeys_Transaction(con);
```

Comment:
Have you tried to run a SELECT FOR UPDATE query on the whole table in mysql and postgresql and seen if there is locking being done?

Reply:
SELECT FOR UPDATE locks the rows that are indexed (indexed is a wrong word here but for lack of better word, I mean rows that the query would traverse and determine if the row needs to be returned) from a query but doesnt lock the entire table, in theory since SELECT * would "touch" all the rows so it would work. However if the table is empty when the query runs (the first time we fetch keys, in the future if the user deletes all keys etc) then it wouldnt lock anything. Still reading up on this

Reply:
Your last comment is untrue based on our experimentation and tests :) (Just keeping this comment here in case we read the above in the future..)